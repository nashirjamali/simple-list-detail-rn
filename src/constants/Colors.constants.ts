export const COLORS = {
  BRAND: '#FB6544',
  BLACK: '#121212',
  SUCCESS: '#55B585',
  GRAY: '#999999',
  WHITE: '#FFFFFF',
} as const;
