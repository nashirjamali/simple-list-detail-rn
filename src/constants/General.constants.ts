export const API_URL =
  'https://recruitment-test.flip.id/frontend-test' as const;
