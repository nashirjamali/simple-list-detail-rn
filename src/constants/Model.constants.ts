export const TRANSACTION_STATUS = {
  SUCCESS: 'SUCCESS',
  PENDING: 'PENDING',
} as const;
