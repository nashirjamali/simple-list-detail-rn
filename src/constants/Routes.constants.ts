export const ROUTES = {
  TRANSACTION_LIST: 'transactionsList',
  TRANSACTION_DETAILS: 'transactionsDetails',
} as const;
