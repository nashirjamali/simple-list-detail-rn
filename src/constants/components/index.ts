export * from './GeneralText.constants';
export * from './Button.constants';
export * from './Label.constants';
export * from './Card.constants';
