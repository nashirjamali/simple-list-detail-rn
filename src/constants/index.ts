export * from './components';
export * from './Colors.constants';
