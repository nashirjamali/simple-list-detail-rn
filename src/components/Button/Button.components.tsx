import React, {FC, ReactElement, ReactNode} from 'react';
import {TouchableOpacity} from 'react-native';

import type {ButtonProps} from '../../@types';
import {BUTTON_VARIANTS, GENERAL_TEXT_VARIANTS} from '../../constants';

import GeneralText from '../GeneralText/GeneralText.components';

import styles from './Button.styles';

/**
 * Renders the title of the button using the GeneralText component.
 *
 * @param {string} title - The text to be displayed as the button title.
 * @returns {ReactElement} The rendered GeneralText component containing the title.
 */
const _renderTitle = (title: string): ReactElement => (
  <GeneralText variant={GENERAL_TEXT_VARIANTS.BUTTON}>{title}</GeneralText>
);

/**
 * Renders an icon if provided.
 *
 * @param {ReactNode} Icon - The icon to be rendered.
 * @returns {ReactNode} The rendered icon or null if not provided.
 */
const _renderIcon = (Icon?: ReactNode): ReactNode => Icon;

/**
 * A customizable button component that can display a label and an icon.
 *
 * @param {ButtonProps} props - The properties for the Button component.
 * @param {string} props.title - The text label to display on the button.
 * @param {string} props.variant - The variant type for the button, determining its style.
 * @param {JSX.Element} [props.iconRight] - An optional icon to display on the right side of the button.
 * @param {JSX.Element} [props.icon] - An optional icon to display on the left side of the button.
 * @returns {JSX.Element} The rendered button component.
 */
const Button: FC<ButtonProps> = ({
  title,
  variant,
  iconRight,
  icon,
  ...props
}: ButtonProps): JSX.Element => (
  <TouchableOpacity style={styles.container} {...props}>
    {variant === BUTTON_VARIANTS.ICON ? _renderIcon(icon) : _renderTitle(title)}
    {_renderIcon(iconRight)}
  </TouchableOpacity>
);

export default Button;
