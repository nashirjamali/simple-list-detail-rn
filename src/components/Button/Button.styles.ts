import {StyleSheet} from 'react-native';

import {COLORS} from '../../constants';
import type {ButtonStyles} from '../../@types';

const {WHITE} = COLORS;

const styles: ButtonStyles = StyleSheet.create({
  container: {
    flexDirection: 'row',
    gap: 2,
    backgroundColor: WHITE,
    justifyContent: 'space-between',
    alignItems: 'center',
  },
});

export default styles;
