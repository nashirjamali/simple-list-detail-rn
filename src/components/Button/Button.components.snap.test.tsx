import React from 'react';
import {render} from '@testing-library/react-native';
import Button from './Button.components';
import {BUTTON_VARIANTS} from '../../constants';

describe('Button Component', () => {
  it('renders correctly with title', () => {
    const {toJSON} = render(
      <Button title="Click Me" variant={BUTTON_VARIANTS.NORMAL} />,
    );
    expect(toJSON()).toMatchSnapshot();
  });

  it('renders correctly with left icon', () => {
    const {toJSON} = render(
      <Button title="Click Me" variant={BUTTON_VARIANTS.NORMAL} icon={<></>} />,
    );
    expect(toJSON()).toMatchSnapshot();
  });

  it('renders correctly with right icon', () => {
    const {toJSON} = render(
      <Button
        title="Click Me"
        variant={BUTTON_VARIANTS.NORMAL}
        iconRight={<></>}
      />,
    );
    expect(toJSON()).toMatchSnapshot();
  });

  it('renders correctly as an icon button', () => {
    const {toJSON} = render(
      <Button
        variant={BUTTON_VARIANTS.ICON}
        icon={<></>}
        title="Icon Button"
      />,
    );
    expect(toJSON()).toMatchSnapshot();
  });
});
