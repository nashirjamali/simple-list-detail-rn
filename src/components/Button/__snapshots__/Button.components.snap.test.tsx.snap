// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button Component renders correctly as an icon button 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={false}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "alignItems": "center",
      "backgroundColor": "#FFFFFF",
      "flexDirection": "row",
      "gap": 2,
      "justifyContent": "space-between",
      "opacity": 1,
    }
  }
/>
`;

exports[`Button Component renders correctly with left icon 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={false}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "alignItems": "center",
      "backgroundColor": "#FFFFFF",
      "flexDirection": "row",
      "gap": 2,
      "justifyContent": "space-between",
      "opacity": 1,
    }
  }
>
  <Text
    style={
      {
        "color": "#FB6544",
        "fontSize": 14,
      }
    }
  >
    Click Me
  </Text>
</View>
`;

exports[`Button Component renders correctly with right icon 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={false}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "alignItems": "center",
      "backgroundColor": "#FFFFFF",
      "flexDirection": "row",
      "gap": 2,
      "justifyContent": "space-between",
      "opacity": 1,
    }
  }
>
  <Text
    style={
      {
        "color": "#FB6544",
        "fontSize": 14,
      }
    }
  >
    Click Me
  </Text>
</View>
`;

exports[`Button Component renders correctly with title 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={false}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "alignItems": "center",
      "backgroundColor": "#FFFFFF",
      "flexDirection": "row",
      "gap": 2,
      "justifyContent": "space-between",
      "opacity": 1,
    }
  }
>
  <Text
    style={
      {
        "color": "#FB6544",
        "fontSize": 14,
      }
    }
  >
    Click Me
  </Text>
</View>
`;
